File: KotlinVersion/app/src/main/java/com/thesis/kotlinversion/MovieAdapter.java
package com.thesis.kotlinversion;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;
import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;
import com.thesis.kotlinversion.database.RoomDB;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.ViewHolder> implements Filterable {

    private List<Movie> movieList;
    private List<Movie> filteredMovieList;
    private Activity context;
    private RecyclerViewClickListener listener;
    private RoomDB database;

    public MovieAdapter(Activity context, List<Movie> movieList, RecyclerViewClickListener listener, RoomDB database) {
        this.movieList = movieList;
        this.listener = listener;
        this.context = context;
        this.filteredMovieList = new ArrayList<>(movieList);
        this.database = database;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context)
                .inflate(R.layout.movie_row, parent, false);
        return new ViewHolder(view);
    }

    @SuppressLint("NotifyDataSetChanged")
    @RequiresApi(api = Build.VERSION_CODES.N) //Req android 7.0 or higher
    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Movie movie = movieList.get(position);

        String movieUrl = movie.getImgUrl();
        if(Utils.isNumeric(movieUrl)) { // Drawable no_img object number
            holder.img.setImageResource(R.drawable.no_img);
        } else {
            Picasso.with(context).load(movie.getImgUrl()).into(holder.img, new Callback() {
                @Override
                public void onSuccess() { }

                @Override
                public void onError() {
                    // If no valid image add no_img from drawable
                    holder.img.setImageResource(R.drawable.no_img);
                }
            });
        }

        String title = movie.getTitle() + " (" + movie.getReleaseDate() + ")";
        holder.title.setText(title);
        holder.plot.setText(movie.getPlot());
        String score = "Score : " + getStars( movie.getScore()) + " " +  movie.getScore() + "/10";
        holder.score.setText(score);

        Chip chip;
        holder.genresLayout.removeAllViews();
        for (String genre:  movie.getGenre()) {
            chip = new Chip(holder.genresLayout.getContext());
            chip.setText(genre);
            chip.setChipBackgroundColor(ColorStateList.valueOf(getRandomColor()));

            chip.setOnClickListener(view -> {
                String chosenGenre = ((Chip) view).getText().toString();
                List<Movie> moviesMatchingGenre = filteredMovieList.stream()
                        .filter(x -> x.getGenre().contains(chosenGenre))
                        .collect(Collectors.toList());

                movieList.clear();
                movieList.addAll( moviesMatchingGenre );
                notifyDataSetChanged();
            });

            holder.genresLayout.addView(chip);
        }

        holder.deleteBtn.setOnClickListener(view -> {
            new AlertDialog.Builder(context)
                    .setTitle(R.string.delSpecDb)
                    .setMessage(R.string.delSpecDbSure)

                    .setPositiveButton(android.R.string.yes, (dialog, which) -> {
                        database.movieDao().delete(movieList.get(position));
                        movieList = database.movieDao().getAll();
                        notifyDataSetChanged();
                        ((Activity)context).recreate(); // Reset start state
                    })

                    .setNegativeButton(android.R.string.no, null)
                    .setIcon(android.R.drawable.ic_dialog_alert)
                    .show();
        });

    }


    public  String getStars(int score) {
        StringBuilder stars = new StringBuilder();
        for (int i = 0; i < score; i++) {
            stars.append("\u2b50 ");
        }
        return stars.toString().trim();
    }

    // minimumBrightnessRatio between 0 and 1. 1 = all white, 0 = any possible color
    private int getRandomColor() {
        Random random = new Random();
        float minimumBrightnessRatio = 0.5F;
        int minimumColorValue = (int) (255f * minimumBrightnessRatio);
        int variableColorValue = 255 - minimumColorValue;

        int r = minimumColorValue + random.nextInt(variableColorValue);
        int g = minimumColorValue + random.nextInt(variableColorValue);
        int b = minimumColorValue + random.nextInt(variableColorValue);

        return Color.argb(255, r, g, b);
    }

    @Override
    public int getItemCount() {
        return movieList.size();
    }

    @Override
    public Filter getFilter() {

        return movieFilter;
    }

    private Filter movieFilter = new Filter() {

        @RequiresApi(api = Build.VERSION_CODES.N) //Req android 7.0 or higher
        @Override
        protected FilterResults performFiltering(CharSequence charSequence) {

            List<Movie> filteredList;
            if(charSequence == null || charSequence.length() == 0) {
                filteredList = new ArrayList<>(filteredMovieList);
            } else {
                String filterPattern = charSequence.toString().trim().toLowerCase();

                filteredList = filteredMovieList.stream()
                        .filter(x ->
                                    x.getTitle().toLowerCase().trim().contains(filterPattern)
                                ||  x.getReleaseDate().equals(filterPattern)
                                ||  x.getPlot().toLowerCase().trim().contains(filterPattern)
                        )
                        .collect(Collectors.toList());
            }

            FilterResults results = new FilterResults();
            results.values = filteredList;

            return results;
        }

        @SuppressLint("NotifyDataSetChanged")
        @Override
        protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
            movieList.clear();
            movieList.addAll( (List) filterResults.values);
            notifyDataSetChanged();
        }
    };


    public class ViewHolder extends RecyclerView.ViewHolder {
        ImageView img;
        TextView title, plot, score;
        ChipGroup genresLayout;
        ImageButton deleteBtn;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            img = itemView.findViewById(R.id.rv_img_view);
            title = itemView.findViewById(R.id.rv_title_view);
            plot = itemView.findViewById(R.id.rv_plot_view);
            score = itemView.findViewById(R.id.rv_score_view);
            genresLayout = (ChipGroup) itemView.findViewById(R.id.chip_group);
            deleteBtn = itemView.findViewById(R.id.rv_delete_btn);
        }
    }
    public interface RecyclerViewClickListener {
        void onClick(View v, int position);
    }

}

File: KotlinVersion/app/src/main/java/com/thesis/kotlinversion/MovieAdapter.kt
package com.thesis.kotlinversion

import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.content.DialogInterface
import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Build
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.annotation.RequiresApi
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso
import com.thesis.kotlinversion.Utils.isNumeric
import com.thesis.kotlinversion.database.RoomDB
import java.util.*
import java.util.stream.Collectors

class MovieAdapter(
    context: Activity,
    movieList: MutableList<Movie?>,
    listener: RecyclerViewClickListener,
    database: RoomDB
) : RecyclerView.Adapter<MovieAdapter.ViewHolder>(), Filterable {
    private var movieList: MutableList<Movie?>?
    private var filteredMovieList: MutableList<Movie?>? = ArrayList(movieList)
    private val context: Activity
    private val listener: RecyclerViewClickListener
    private val database: RoomDB
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(context)
            .inflate(R.layout.movie_row, parent, false)
        return ViewHolder(view)
    }

    init {
        this.movieList = movieList
        this.listener = listener
        this.context = context
        this.database = database
    }

    @SuppressLint("NotifyDataSetChanged")
    @RequiresApi(api = Build.VERSION_CODES.N) //Req android 7.0 or higher
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val movie = movieList!![position]
        val movieUrl = movie!!.imgUrl
        if (isNumeric(movieUrl)) { // Drawable no_img object number
            holder.img.setImageResource(R.drawable.no_img)
        } else {
            Picasso.with(context).load(movie.imgUrl).into(holder.img, object : Callback {
                override fun onSuccess() {}
                override fun onError() {
                    // If no valid image add no_img from drawable
                    holder.img.setImageResource(R.drawable.no_img)
                }
            })
        }
        val title = movie.title + " (" + movie.releaseDate + ")"
        holder.title.text = title
        holder.plot.text = movie.plot
        val score = "Score : " + getStars(movie.score) + " " + movie.score + "/10"
        holder.score.text = score
        var chip: Chip
        holder.genresLayout.removeAllViews()
        for (genre in movie.genre!!) {
            chip = Chip(holder.genresLayout.context)
            chip.text = genre
            chip.chipBackgroundColor = ColorStateList.valueOf(randomColor)
            chip.setOnClickListener { view: View ->
                val chosenGenre = (view as Chip).text.toString()
                val moviesMatchingGenre = filteredMovieList!!.stream()
                    .filter { x: Movie? -> x!!.genre!!.contains(chosenGenre) }
                    .collect(Collectors.toList())
                movieList!!.clear()
                movieList!!.addAll(moviesMatchingGenre)
                notifyDataSetChanged()
            }
            holder.genresLayout.addView(chip)
        }
        holder.deleteBtn.setOnClickListener { view: View? ->
            AlertDialog.Builder(context)
                .setTitle(R.string.delSpecDb)
                .setMessage(R.string.delSpecDbSure)
                .setPositiveButton("yes") { dialog: DialogInterface?, which: Int ->
                    database.movieDao()!!.delete(movieList!![position])
                    movieList = database.movieDao()!!.all
                    notifyDataSetChanged()
                    context.recreate() // Reset start state
                }
                .setNegativeButton("no", null)
                .setIcon(android.R.drawable.ic_dialog_alert)
                .show()
        }
    }

    private fun getStars(score: Int): String {
        val stars = StringBuilder()
        for (i in 0 until score) {
            stars.append("\u2b50 ")
        }
        return stars.toString().trim { it <= ' ' }
    }

    // minimumBrightnessRatio between 0 and 1. 1 = all white, 0 = any possible color
    private val randomColor: Int
        get() {
            val random = Random()
            val minimumBrightnessRatio = 0.5f
            val minimumColorValue = (255f * minimumBrightnessRatio).toInt()
            val variableColorValue = 255 - minimumColorValue
            val r = minimumColorValue + random.nextInt(variableColorValue)
            val g = minimumColorValue + random.nextInt(variableColorValue)
            val b = minimumColorValue + random.nextInt(variableColorValue)
            return Color.argb(255, r, g, b)
        }

    override fun getItemCount(): Int {
        return movieList!!.size
    }

    override fun getFilter(): Filter {
        return movieFilter
    }

    private val movieFilter: Filter = object : Filter() {
        @RequiresApi(api = Build.VERSION_CODES.N) //Req android 7.0 or higher
        override fun performFiltering(charSequence: CharSequence): FilterResults {
            var filteredList: MutableList<Movie>? = ArrayList()
                if (charSequence.isEmpty()) {
                    filteredList?.addAll(filteredMovieList as Collection<Movie>)
            } else {
                val filterPattern = charSequence.toString().trim { it <= ' ' }.lowercase()

                    filteredList = filteredMovieList?.filter { x: Movie? ->
                        (x!!.title!!.lowercase().trim { it <= ' ' }
                            .contains(filterPattern)
                                || x.releaseDate == filterPattern || x.plot!!.lowercase()
                            .trim { it <= ' ' }
                            .contains(filterPattern))
                    } as MutableList<Movie>?

            }
            val results = FilterResults()
            results.values = filteredList
            return results
        }

        @SuppressLint("NotifyDataSetChanged")
        override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {
            movieList.clear()
            movieList.addAll(filterResults.values as Collection<Movie?>)
            notifyDataSetChanged()
        }
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var img: ImageView = itemView.findViewById(R.id.rv_img_view)
        var title: TextView = itemView.findViewById(R.id.rv_title_view)
        var plot: TextView = itemView.findViewById(R.id.rv_plot_view)
        var score: TextView = itemView.findViewById(R.id.rv_score_view)
        var genresLayout: ChipGroup = itemView.findViewById<View>(R.id.chip_group) as ChipGroup
        var deleteBtn: ImageButton = itemView.findViewById(R.id.rv_delete_btn)

    }

    interface RecyclerViewClickListener {
        fun onClick(v: View?, position: Int)
    }
}